OPTIMADE JSON Lines specification appendix (#531)
Materials-Consortia/OPTIMADE: optimade.rst

- The next line MAY contain a standard OPTIMADE ``meta`` object, following the same rules described in `JSON Response Schema: Common Fields`_, where every MUST and SHOULD rule can be reinterpreted as a MAY rule.
- The next block of lines provides the ``info`` endpoint responses. 
  - First the base info response MUST be provided, following the description at `Base Info Endpoint`_.
  - The next lines MUST contain the entry info endpoint responses for the all entry types present later in the file, as described in `Entry Listing Info Endpoints`_. 
    These MUST appear in alphabetical order by entry type name.
- The remaining lines of the file contain data entries themselves, described in `Entry Listng JSON Response Schema`_.
  Again, these MUST appear in block alphabetical order by entry type name, but can appear in any order within those blocks.
- Finally, any custom extension endpoints (see `Custom Extension Endpoints`_), if present and desirable, MUST appear at the end of the file.

This leaves the following overall file structure:

> .. code:: jsonc

Author: ```suggestion
.. code
```

> ```
> <header>
> <optional metadata>
> <base info response>
> <entry info responses>
> <entries block ordered by entry type>
> <optional custom extension endpoints>
> ```

Author: ```suggestion
.. code:: jsonc
    <header>
    <optional metadata>
    <base info response>
    <entry info responses>
    <entries block ordered by entry type>
    <optional custom extension endpoints>

```

> .. code
>     <header>
>     <optional metadata>
>     <base info response>
>     <entry info responses>
>     <entries block ordered by entry type>
>     <optional custom extension endpoints>

Author: ```suggestion
.. code
  <header>
  <optional metadata>
  <base info response>
  <entry info responses>
  <entries block ordered by entry type>
  <optional custom extension endpoints>
```

- Information about prefixes and provider fields

Example OPTIMADE JSON Lines File
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> .. code:jsonc

Author: ```suggestion
.. code:: jsonc
```

> .. code:: jsonc

Author: ```suggestion
.. code
```

> ```jsonlines
> {"x-optimade": {"api_version": "1.2.0"}}
> {"meta": {"time_stamp": "2024-07-19T11:47:10Z", "data_returned": 6, "provider": {"name": "Example JSONL", "description": "An example JSONL file.", "prefix": "_exmpl"}}}
> {"type": "info", "id": "/", "attributes": {"api_version": "1.2.0", "available_api_versions": ["1.2.0"], "formats": ["json"], "entry_types_by_format": {"json": ["references", "structures"]}, "license": "https://example.com/licenses/example_license.html"} "homepage": "https://example.com", "name": "Example API", "provider": {"description": "A simple example provider", "name": "Example Provider"}}}
> {"type": "info", "id": "references", ...}
> {"type": "info", "id": "structures", ...}
> {"type": "structures", "id": "1", "attributes": {...}
> ```

Author: ```suggestion
.. code:jsonc
    {"x-optimade": {"api_version": "1.2.0"}}
    {"meta": {"time_stamp": "2024-07-19T11:47:10Z", "data_returned": 6, "provider": {"name": "Example JSONL", "description": "An example JSONL file.", "prefix": "_exmpl"}}}
    {"type": "info", "id": "/", "attributes": {"api_version": "1.2.0", "available_api_versions": ["1.2.0"], "formats": ["json"], "entry_types_by_format": {"json": ["references", "structures"]}, "license": "https://example.com/licenses/example_license.html"} "homepage": "https://example.com", "name": "Example API", "provider": {"description": "A simple example provider", "name": "Example Provider"}}}
    {"type": "info", "id": "references", ...}
    {"type": "info", "id": "structures", ...}
    {"type": "structures", "id": "1", "attributes": {...}

```

> .. code
>     {"x-optimade": {"api_version": "1.2.0"}}
>     {"meta": {"time_stamp": "2024-07-19T11:47:10Z", "data_returned": 6, "provider": {"name": "Example JSONL", "description": "An example JSONL file.", "prefix": "_exmpl"}}}
>     {"type": "info", "id": "/", "attributes": {"api_version": "1.2.0", "available_api_versions": ["1.2.0"], "formats": ["json"], "entry_types_by_format": {"json": ["references", "structures"]}, "license": "https://example.com/licenses/example_license.html"} "homepage": "https://example.com", "name": "Example API", "provider": {"description": "A simple example provider", "name": "Example Provider"}}}
>     {"type": "info", "id": "references", ...}
>     {"type": "info", "id": "structures", ...}
>     {"type": "structures", "id": "1", "attributes": {...}

Author: ```suggestion
.. code
  {"x-optimade": {"api_version": "1.2.0"}}
  {"meta": {"time_stamp": "2024-07-19T11:47:10Z", "data_returned": 6, "provider": {"name": "Example JSONL", "description": "An example JSONL file.", "prefix": "_exmpl"}}}
  {"type": "info", "id": "/", "attributes": {"api_version": "1.2.0", "available_api_versions": ["1.2.0"], "formats": ["json"], "entry_types_by_format": {"json": ["references", "structures"]}, "license": "https://example.com/licenses/example_license.html"} "homepage": "https://example.com", "name": "Example API", "provider": {"description": "A simple example provider", "name": "Example Provider"}}}
  {"type": "info", "id": "references", ...}
  {"type": "info", "id": "structures", ...}
  {"type": "structures", "id": "1", "attributes": {...}
```

    * XML Schema appears to use a compatible regex format, except it is implicitly anchored: i.e., the beginning-of-input ``^`` and end-of-input ``$`` anchors must be removed, and missing anchors replaced by ``.*``.
    * POSIX Extended regexes (and their extended GNU implementations) are incompatible because ``\`` is not a special character in character classes.
      POSIX Basic regexes also have further differences, e.g., the meaning of some escaped syntax characters is reversed.

The OPTIMADE JSON Lines Format for Database Exchange
----------------------------------------------------

There are many use cases for which it is beneficial to share all of the data served by an OPTIMADE API as a single file, for example, archival, transfer of entire databases and local-first clients.
This appendix describes a lightweight standardization for doing this via the `JSON Lines <https://jsonlines.org/>`__ format, with some additional OPTIMADE-specific conventions.

The `JSON Lines <https://jsonlines.org/>`__ format enforces the following rules:

> - each line is a valid JSON object,

Author: Hmm, not sure, to me value implies something stored under a key, whereas object is the smallest chunk of data that can be parsed entirely as JSON?

> - each line is a valid JSON object,

Reviewer: on https://jsonlines.org/ they mention "values" (as even e.g. just strings or integers are supported). But I agree that in the vast majority of cases, people will use objects.

- each line is separated by a newline character (``\n``), optionally ending the file with a newline.
- each file must be UTF-8 encoded,
- the recommended file extension is ``.jsonl``, with natural extensions to ``.jsonl.gz`` and ``jsonl.bz2`` for ``gzip`` and ``bzpi2`` compressed files, respectively.

The OPTIMADE JSON Lines format then extends these rules with the following conventions:

- The first line of the file is a JSON object that contains metadata about the file.
  It MUST comprise of a dictionary with the key ``x-optimade``, under which the following key MUST be defined:

  - ``api_version``: The OPTIMADE API version used when generating the file, as described in the ``meta`` member in `JSON Response Schema: Common Fields`_.

- The next line MAY contain a standard OPTIMADE ``meta`` object, following the same rules described in `JSON Response Schema: Common Fields`_, where every MUST and SHOULD rule can be reinterpreted as a MAY rule.
- The next block of lines provides the ``info`` endpoint responses.

>   - First the base info response MUST be provided, following the description at `Base Info Endpoint`_.

Reviewer: > MUST

This breaks current optimade-maker jsonl files (well, just the li-ion-conductors one). Would a MAY, in principle, work here?

Author: This is quite important IMO, we don't provide any other way of identifying the file (and in fact we might want to even expand this with a version of the JSONL specification like ``{'x-optimade': {'api_version': 1.1, 'jsonl_version': 1.2}`` if we want the JSONL file to be able to store OPTIMADE data of any version. Would you be against simply adding it to the files where it is missing atm?

Reviewer: Sorry, not sure if my comment was precise enough. I was commenting on the base info endpoint, which doesn't contain any crucial info as far as i understand, but I see how it can be useful to know what follows in the next lines.

>   - The next lines MUST contain the entry info endpoint responses for the all entry types present later in the file, as described in `Entry Listing Info Endpoints`_. These MUST appear in alphabetical order by entry type name.

Reviewer: > These MUST appear in alphabetical order by entry type name.

Fine for me, but is it needed?

Author: Happy to relax this constraint, probably we'll never need to worry about an OPTIMADE JSONL file containing 100s of entry types (in which case you could use alphabetical order to short-circuit the search for a given entry type).

Reviewer: ok, i see, but anyway, fine for me. good to keep consistent if we require the  entry blocks to be sorted, where this is more valuable, i think.

Author: Revisiting this, I'm going to remove the constraint and allow entry types to be interleaved. I've found this useful when scraping another database and pulling included entries, for large databases this would require an additional final sorting step which is not ideal.

These should be consistent with any external definitions, and where appropriate, prefixes tied to the tools used to generate the file should be used.
It is RECOMMENDED that custom properties are defined in full within the JSONL file, or pointed to a specific versioned property definition.

Example OPTIMADE JSON Lines File
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code :: jsonc

  {"x-optimade": {"api_version": "1.2.0"}}
  {"meta": {"time_stamp": "2024-07-19T11:47:10Z", "data_returned": 6, "provider": {"name": "Example JSONL", "description": "An example JSONL file.", "prefix": "_exmpl"}}}
  {"type": "info", "id": "/", "attributes": {"api_version": "1.2.0", "available_api_versions": ["1.2.0"], "formats": ["json"], "entry_types_by_format": {"json": ["references", "structures"]}, "license": "https://example.com/licenses/example_license.html"}, "homepage": "https://example.com", "name": "Example API", "provider": {"description": "A simple example provider", "name": "Example Provider"}}}
  {"type": "info", "id": "references", ...}
  {"type": "info", "id": "structures", ...}
  {"type": "structures", "id": "1", "attributes": {...}}

>   {"type": "references", "id": "2", "attributes": {...}}

Reviewer: `references` block should be ahead of `structures`, as required above. But again, is the alphabetical ordering needed?

>   {"type": "structures", "id": "1", "attributes": {...}}
>   {"type": "references", "id": "2", "attributes": {...}}

Reviewer: ```suggestion
  {"type": "references", "id": "1", "attributes": {...}}
  {"type": "structures", "id": "2", "attributes": {...}}
```

switched em

    * XML Schema appears to use a compatible regex format, except it is implicitly anchored: i.e., the beginning-of-input ``^`` and end-of-input ``$`` anchors must be removed, and missing anchors replaced by ``.*``.
    * POSIX Extended regexes (and their extended GNU implementations) are incompatible because ``\`` is not a special character in character classes.
      POSIX Basic regexes also have further differences, e.g., the meaning of some escaped syntax characters is reversed.

The OPTIMADE JSON Lines Format for Database Exchange
----------------------------------------------------

There are many use cases for which it is beneficial to share all of the data served by an OPTIMADE API as a single file, for example, archival, transfer of entire databases and local-first clients.

> This appendix describes a lightweight standardization for doing this via the `JSON Lines <https://jsonlines.org/>`__ format, with some additional OPTIMADE-specific conventions.

Reviewer 2: ```suggestion
This appendix describes a lightweight standardization for doing this via the `JSON Lines <https://jsonlines.org/>`__ (JSONL) format, with some additional OPTIMADE-specific conventions.
```
The following text uses JSONL abbreviation, I suggest introducing it here.

The `JSON Lines <https://jsonlines.org/>`__ format enforces the following rules:

- each line is a valid JSON value,
- each line is separated by a newline character (``\n``), optionally ending the file with a newline.
- each file must be UTF-8 encoded,
- the recommended file extension is ``.jsonl``, with natural extensions to ``.jsonl.gz`` and ``jsonl.bz2`` for ``gzip`` and ``bzip2`` compressed files, respectively.

The OPTIMADE JSON Lines format then extends these rules with the following conventions:

- The first line of the file is a JSON object that contains metadata about the file.
  It MUST be a dictionary with the key ``x-optimade``, under which the following key MUST be defined:

  - ``api_version``: The OPTIMADE API version used when generating the file, as described in the ``meta`` member in `JSON Response Schema: Common Fields`_.

> - The next line MAY contain a standard OPTIMADE ``meta`` object, following the same rules described in `JSON Response Schema: Common Fields`_, where every MUST and SHOULD rule can be reinterpreted as a MAY rule.

Reviewer: Should we specify what meaning fields like `meta.data_returned` take in this context?

Author: ```suggestion
- The next line MAY contain a standard OPTIMADE ``meta`` object, following the same rules described in `JSON Response Schema: Common Fields`_, where every MUST and SHOULD rule can be reinterpreted as a MAY rule. For example, `meta.data_returned` can be omitted, or otherwise used to provide the total number of entries in the file.
```

Something like this?

Reviewer: @ml-evs ok, yep, sounds good.

- The next block of lines provides the ``info`` endpoint responses.

>   - First the base info response MUST be provided, following the description at `Base Info Endpoint`_.

Reviewer: this sublist is not rendering correctly to me

Author: fixed, thanks

  - The next lines MUST contain the entry info endpoint responses for the all entry types present later in the file, as described in `Entry Listing Info Endpoints`_.
- The remaining lines of the file contain data entries themselves, described in `Entry Listing JSON Response Schema`_.

>   These entries can be provided in any order.

Reviewer: The order of the data entries is not clear to me. Below, we say `<entries block ordered by entry type>`, while here they can be in any order. I don't have strong feelings about this, but it should be clear. If we want to mandate block-order (e.g. all structures first, then all references, etc), i guess we should say something about both, 1) the order of the blocks (should they match the info endpoint order?); 2) the order of the entries within the blocks.

My own gut feeling is that perhaps the order of blocks is not important, but maybe it's good to have one type of entries all together in a block.

- Finally, any custom extension endpoints (see `Custom Extension Endpoints`_), if present and desirable, MUST appear at the end of the file.

This leaves the following overall file structure:

.. code :: txt

  <header>
  <optional metadata>
  <base info response>
  <entry info responses>
  <entries block ordered by entry type>
  <optional custom extension endpoints>

This JSONL format can also be used to share provider-specific properties.

> These should be consistent with any external definitions, and where appropriate, prefixes tied to the tools used to generate the file should be used.

Reviewer 2: I do not quite get the "prefixes tied to the tools used to generate the file should be used" part. Are there any examples for such cases?

Author: Hmm, yeah, this has lost its meaning a bit. The intention was that a tool like aiida or ASE could support OPTIMADE JSONL as an export format, and any "tool-specific definitions" like `_aiida_node_id` (or whatever) should be defined and use the tool prefix. I wouldn't be against removing this sentence as the format just follows the normal OPTIMADE rules about defining properties.

Reviewer 2: Right, thanks for the explanation. Since "MUST" is not used in this paragraph, I gather its main purpose is to recommend best practice. We may probably want that all the OPTIMADE JSONL files bear some sort of producer identification, but we probably do not want to standardize it at this point. But maybe we can reuse `meta.implementation` dictionary from the main specification?

Maybe it is worth stressing that OPTIMADE JSONL files have the same requirements ("MUST/SHOULD/MAY") as OPTIMADE responses, in terms of custom properties having prefixes and metadata descriptions. But maybe it is enough just to say that?

Also, what about sharing provider-specific entry types? If we mention provider-specific properties, should be allow entry types as well?

Reviewer: Regarding non-optimade-spec properties, doesn't the spec say they must have either

- the provider specific prefix; or
- a defininition-provider-specific prefix?

I think we can just point to section 3.5 here, no?

It is RECOMMENDED that custom properties are defined in full within the JSONL file, or pointed to a specific versioned property definition.

> Example OPTIMADE JSON Lines File

Reviewer: ```suggestion
Example file
```
Considering there is now two `OPTIMADE JSON Lines File` standards (other one being 9.4), maybe better to be more vague here (e.g. i already mixed these up when looking through the table of contents).

This appendix describes a lightweight standardization for doing this via the `JSON Lines <https://jsonlines.org/>`__ (JSONL) format, with some additional OPTIMADE-specific conventions.

The `JSON Lines <https://jsonlines.org/>`__ format enforces the following rules:

- each line is a valid JSON value,
- each line is separated by a newline character (``\n``), optionally ending the file with a newline.
- each file must be UTF-8 encoded,
- the recommended file extension is ``.jsonl``, with natural extensions to ``.jsonl.gz`` and ``jsonl.bz2`` for ``gzip`` and ``bzip2`` compressed files, respectively.

The OPTIMADE JSON Lines format then extends these rules with the following conventions:

- The first line of the file is a JSON object that contains metadata about the file.
  It MUST be a dictionary with the key ``x-optimade``, under which the following key MUST be defined:

  - ``api_version``: The OPTIMADE API version used when generating the file, as described in the ``meta`` member in `JSON Response Schema: Common Fields`_.

> - The next line MAY contain a standard OPTIMADE ``meta`` object, following the same rules described in `JSON Response Schema: Common Fields`_, where every MUST and SHOULD rule can be reinterpreted as a MAY rule. For example, `meta.data_returned` can be omitted, or otherwise used to provide the total number of entries in the file.

Reviewer: ```suggestion
- The next line MAY contain a standard OPTIMADE ``meta`` object, following the same rules described in `JSON Response Schema: Common Fields`_, where every MUST and SHOULD rule can be reinterpreted as a MAY rule. For example, ``meta.data_returned`` can be omitted, or otherwise used to provide the total number of entries in the file.
```

> - The next line MAY contain a standard OPTIMADE ``meta`` object, following the same rules described in `JSON Response Schema: Common Fields`_, where every MUST and SHOULD rule can be reinterpreted as a MAY rule. For example, ``meta.data_returned`` can be omitted, or otherwise used to provide the total number of entries in the file.

Reviewer 3: ```suggestion
- The next line MAY contain a standard OPTIMADE ``meta`` object, following the same rules described in `JSON Response Schema: Common Fields`_, where every MUST and SHOULD rule can be reinterpreted as a MAY rule.
  For example, ``meta.data_returned`` can be omitted, or otherwise used to provide the total number of entries in the file.
```

Reviewer 3: Feel free to dismiss this comment if you would rather have this formatting change made after the merge.

  - First the base info response MUST be provided, following the description at `Base Info Endpoint`_.

  - The next lines MUST contain the entry info endpoint responses for the all entry types present later in the file, as described in `Entry Listing Info Endpoints`_.

- The remaining lines of the file contain data entries themselves, described in `Entry Listing JSON Response Schema`_.
  These entries can be provided in any order.

- Finally, any custom extension endpoints (see `Custom Extension Endpoints`_), if present and desirable, MUST appear at the end of the file.

This leaves the following overall file structure:

.. code :: txt

  <header>
  <optional metadata>
  <base info response>
  <entry info responses>
  <unordered entries>
  <optional custom extension endpoints>

This JSONL format can also be used to share provider-specific or namespace-specific properties, following the normal rules defined in `Namespace Prefixes`_.
Such static property definitions should be consistent with any externally defined prefixes, that is, JSONLines files should not use an existing prefix unless the desired definitions match exactly.

> Where appropriate, prefixes tied to the tools used to generate the file can be used, in cases where an external property definition is not provided (for example, a property that only applies to the current dataset).

Reviewer: @ml-evs thanks, this seems good. Although, would something like the suggestion below be more precise? Basically saying that there can only be one "custom" tool-based prefix and it should take "role" of the provider prefix in this case.

```suggestion
Where appropriate, a provider prefix can be chosen to represent the tool used to generate the file, in cases where an external property definition is not provided (for example, a property that only applies to the current dataset).
```

Author: Yep, I like that, thanks!

This JSONL format can also be used to share provider-specific or namespace-specific properties, following the normal rules defined in `Namespace Prefixes`_.
Such static property definitions should be consistent with any externally defined prefixes, that is, JSONLines files should not use an existing prefix unless the desired definitions match exactly.
Where appropriate, a provider prefix can be chosen to represent the tool used to generate the file, in cases where an external property definition is not provided (for example, a property that only applies to the current dataset).

It is RECOMMENDED that custom properties are defined in full within the JSONL file, or pointed to a specific versioned property definition.

Example file

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
~~~~~~~~~~~~
```
